macros/src/lib.rs

```diff
@@ -131,10 +89,151 @@ impl DerivedTS {
+        let generic_params = generics
+            .params
+            .iter()
+            .filter(|param| matches!(param, GenericParam::Type(_)))
+            .map(|_| quote! { () });
+        let ty = quote!(<#rust_ty<#(#generic_params),*> as ts_rs::TS>);
+
+        Some(quote! {
+            #[cfg(test)]
+            #[test]
+            fn #test_fn() {
+                #ty::export().expect("could not export type");
+            }
+        })
+    }
+
+    fn generate_name_fn(&self) -> TokenStream {
+        let name = self.name_with_generics();
+        quote! {
+             fn name() -> String {
+                 #name
+             }
+        }
+    }
+
+    fn generate_inline_fn(&self) -> TokenStream {
+        let inline = &self.inline;
+
+        let inline_flattened = self.inline_flattened.as_ref().map(|inline_flattened| {
+            quote! {
+                fn inline_flattened() -> String {
+                    #inline_flattened
+                }
+            }
+        });
+        let inline = quote! {
+            fn inline() -> String {
+                #inline
+            }
+        };
+        quote! {
+            #inline
+            #inline_flattened
+        }
+    }
+
+    /// Generates the `decl()` and `decl_concrete()` methods.
+    /// `decl_concrete()` is simple, and simply defers to `inline()`.
+    /// For `decl()`, however, we need to change out the generic parameters of the type, replacing
+    /// them with the dummy types generated by `generate_generic_types()`.
+    fn generate_decl_fn(&mut self, rust_ty: &Ident) -> TokenStream {
+        let name = &self.ts_name;
+        let generic_types = self.generate_generic_types();
+        let ts_generics = format_generics(&mut self.dependencies, &self.generics);
+        // These are the generic parameters we'll be using.
+        let generic_idents = self.generics.params.iter().filter_map(|p| match p {
+            GenericParam::Lifetime(_) => None,
+            // Since we named our dummy types the same as the generic parameters, we can just keep
+            // the identifier of the generic parameter - its name is shadowed by the dummy struct.
+            GenericParam::Type(TypeParam { ident, .. }) => Some(quote!(#ident)),
+            // We keep const parameters as they are, since there's no sensible default value we can
+            // use instead. This might be something to change in the future.
+            GenericParam::Const(ConstParam { ident, .. }) => Some(quote!(#ident)),
+        });
+        quote! {
+            fn decl_concrete() -> String {
+                format!("type {} = {};", #name, Self::inline())
+            }
+            fn decl() -> String {
+                #generic_types
+                let inline = <#rust_ty<#(#generic_idents,)*> as ts_rs::TS>::inline();
+                let generics = #ts_generics;
+                format!("type {}{generics} = {inline};", #name)
+            }
+        }
+    }
 }
```

Comment:
And finally, `decl` and `decl_concrete` are implemented here.  
`decl` first creates the dummy types, shadowing the names of the generics (`#generic_types`). Then, I just call `inline` on that type.